README.md
# AVTOMAT
Non-deterministic Finite State Machine (with Empty Moves) Implementation

## Usage
Add `<script type="text/javascript" src="path/to/scripts/avtomat-1.0.js">` in
your HTML `<head>` or `<body>`.

## Reference

### From version 1.0
* `.state()` - gets the current state
* `.input(symbol)` - inputs a symbol to the state machine
* `.reset()` - resets the machine
* `.accepted()` - determines if at least one of the current states are final

### Added as of version 1.1 *(See changelog.md)*
* `.nullState()` - determines if the machine is in null state
* `.addState(id, isFinal, transitions)`
* `.deleteState(id)`
* `.addTransition(state_id_from, input_symbol, state_id_to)`
* `.deleteTransition(state_id_from, input_symbol)`
* `.hasTransition(state_id, input_symbol)`
* `.hasTransitions(state_id)`

#### Event related
* `.bindStateEvent(state_id, type, fn)`

`type` should be one of the following:
- `arriving` - when state is becoming a current state
- `arrive` - when state is a current state
- `leaving` - when state is losing its current state status
- `leave` - when state is no longer a current state

* `.bindMachineEvent(type, fn)`

`type` should be one of the following:
- `changing` - when machine is changing state
- `change` - when machine has changed state

## Initialization
The syntax of the constructor is:

    Avtomat.StateMachine(states, startState)

where:
* `states` is the JSON object containing the details of the machine's states (see below)
* `startState` is the ID of the state acting as a start state of the machine

The `states` object has this following structure:

    {
        stateID: {
            "final": <boolean>,
            "transitions": {
                inputSymbol: stateID,
                inputSymbol: stateID,
                .
                .
                .
                inputSymbol: stateID,
            }
        },
        stateID: {
            .
            .
            .
        },
        .
        .
        .
    }
    
where:
* `stateID` is a string which acts as identifier for a state
* `<boolean>` is a Boolean value of `true` or `false`
* `inputSymbol` is a string or a number (probably objects as well)

To initialize an instance of the state machine object:

	var foo = new Avtomat.StateMachine({
		"stateName": {
			"final": false // boolean
			"transitions": {
				// stateSymbol could be integer, float, or string
				"stateSymbol": "anotherStateName",
				"stateSymbol2": ["stateName1", "stateName2"],
				...
				"": "stateName3" // for the empty moves, put an empty string
			}
		}
	}, "stateName2"); // stateName2 is the start state

## Example

	var automaton = new Avtomat.StateMachine({
		"A": {
			"final": false, "transitions": {
				"b": "B",
				"c": ["C", "D"],
				"d": "A",
				"": "E"
			}
		},
		"B": {
			"final": false, "transitions": {
				"a": "A",
				"c": "C",
				"d": "E"
			}
		},
		"C": {
			"final": false,
			"transitions": {
				"": "A"
			}
		},
		"D": {
			"final": true,
			"transitions": {
				"a": "D",
				"b": "E",
				"c": ["B", "D"],
				"d": "D",
				"e": "E"
			}
		},
		"E": {
			"final": false,
			"transitions": {
				"d": "D",
				"c": "C",
			}
		}
	}, "A");

	automaton.state(); // ["E"]: A("") => E
	automaton.input("d"); // ["D"]: E("d") => D
	automaton.accepted(); // true: D is final state
	automaton.input("B"); // []: No such transition, goes to null state
	automaton.input("a"); // []: Null states have no transitions
	automaton.nullState(); // true
	automaton.reset(); // ["E"]: Reset to original state A, A("") => E
	automaton.input("d"); // ["D"]: E("d") => D
	automaton.input("c"); // ["B", "D"]: D("c") => {B, D}
	automaton.input("d"); // ["D", "E"]: B("d") => E, D("d") => D
	automaton.accepted(); // true: D is final state
	automaton.input("e"); // ["E"]: D("e") => E, E("e") => null
	automaton.accepted(); // false: E is not a final state
	automaton.reset(); // ["E"]: Reset to original state A, A("") => E
	automaton.input("c"); // ["E"]: E("c") => C, C("") => A, A("") => E
 


changelog.md
# 1.1
* Added public methods:
	- For states
		- `nullState()`
		- `addState()`
		- `deleteState()`
	- For transitions
		- `addTransition()`
		- `deleteTransition()`
		- `hasTransition()`
		- `hasTransitions()`
	- For event binding
		- `bindStateEvent()`
		- `bindMachineEvent()`
# 1.0
* Initial release

src/avtomat-1.0.js
/**                                                                                                                  *
 *  _______________________________________________________________________________________________________________  *
 *                                                                                                                   *
 *    AVTOMAT - Non-deterministic Finite State Machine (with Empty Moves) Implementation                             *
 *  _______________________________________________________________________________________________________________  *
 *                                                                                                                   *
 *    Version: 1.0                                                                                                   *
 *    Author:  Temoto-kun                                                                                            *
 *    Date:    2013 December 13   08:50 [UTC+8]                                                                      *
 *                                                                                                                   *
 *  =============================================================================================================== **/

(function(w) {
	w.Avtomat = w.Avtomat || {
		/**
		 * A non-deterministic finite state machine which supports empty moves for transitions.
		 * @param states JSON object containing the states, and their details (is final state, transitions).
		 * @param startState The ID of the start state
		 * @returns {Avtomat.StateMachine} New instance of the Avtomat.StateMachine
		 * @constructor
		 */
		StateMachine : function(states, startState) {



			/* =============== *
			 * Private Members *
			 * =============== */

			/* ------ *
			 * Fields *
			 * ------ */

 			var
				// The states
				_states = [],

				// The ID for the "null" state
				_nullStateId = null,

				// The string for the empty input
				_emptyInput = "",

				// IDs of the current state
				_cState = [_nullStateId],

				// ID of the start state
				_startStateId = startState || _nullStateId
				;

			/* ------- *
			 * Methods *
			 * ------- */

 			/**
			 * Adds a transition from one state to another.
			 * @param idFrom The ID of the source state
			 * @param input The input symbol
			 * @param idTo The ID of the destination state
			 * @private
			 */
			function _addTransition(idFrom, input, idTo) {
				_states[idFrom].transitions[input] = idTo;
			}

			/**
			 * Adds a new state.
			 * @param id The ID of the new state
			 * @param isFinal Is state a final state?
			 * @param transitions Associative array containing the input symbol, and the destination state id
			 * @see _createBlankState
			 * @see _addTransition
			 * @private
			 */
			function _addState(id, isFinal, transitions) {
				if(_states[id] === undefined) {
					if(_startStateId == _nullStateId)
						// make the first non-null state to be startState, given that there
						_setStartState(id);

					_states[id] = _createBlankState();

					if(id != _nullStateId) {
						_setFinal(id, isFinal);
						for(var input in transitions) {
							//noinspection JSUnfilteredForInLoop
							_addTransition(id, input, transitions[input]);
						}
					}
				}
			}

			/**
			 * Creates a prefab blank state.
			 * @returns {{final: boolean, transitions: Array}} New blank non-final state with no transitions
			 * @private
			 */
			function _createBlankState() {
				return { final: false, transitions: [] };
			}

			/**
			 * Sets a state to be final state.
			 * @param id The ID of the state
			 * @param bFinal Is state a final state?
			 * @private
			 */
			function _setFinal(id, bFinal) {
				if(id != _nullStateId)
					_states[id].final = bFinal;
			}

			/**
			 * Resets the machine.
			 * @returns {Array} The current state
			 * @private
			 */
			function _reset() {
				_cState = [_startStateId];
				_input("");
				return _cState;
			}

			/**
			 * Inputs a symbol to the machine.
			 * @param input The input symbol
			 * @returns {Array} The resulting state
			 * @private
			 */
			function _input(input) {
				var
					// Queue for old states
					srcStates = _cState,

					// Stack for new states
					newStates = [],

					// If the new states have empty moves
					haveEmptyMoves = false;

				/*
				 * _input operates with a queue as container for the state IDs for processing. It empties srcStates to
				 * make sure all the states will receive the input.
				 */
				while(srcStates.length > 0) {
					// Dequeue
					var oldState = srcStates.shift();

					// Destination states (non-array), i.e. state ID strings by themselves
					var destStateObjs = _states[oldState].transitions[input]
						// if there is no explicity-defined destination states for empty moves, current state is
						// retained
						|| (input == _emptyInput ? oldState : _nullStateId);
					// Destination states (array)
					var destStates;

					// Destination states must be turned into an array for processing
					if(destStateObjs instanceof Array)
						destStates = destStateObjs;
					else
						destStates = [destStateObjs];

					// Determine the destination states
					for(var j = 0; j < destStates.length; j++) {
						var containsState = false;
						for(var k = 0, cStateLen = newStates.length; k < cStateLen; k++)
							containsState = (containsState || (newStates[k] == destStates[j]));
						if(!containsState && destStates[j] != _nullStateId)
							// Make sure the destination states don't duplicate. Null states are not included as well.
							newStates.push(destStates[j]);
					}

					// Terminate the state transitions when there are no explicitly-defined empty transitions
					if(input == _emptyInput && destStateObjs == _nullStateId)
						haveEmptyMoves = false;
				}
				_cState = newStates;

				// can't have empty moves when machine doesn't have current states
				if(newStates.length > 0) {
					for(var l in newStates) { //noinspection JSUnfilteredForInLoop
						if(_states[newStates[l]].transitions[_emptyInput] !== undefined) {
							haveEmptyMoves = true;
							break;
						}
					}
					if(haveEmptyMoves)
						_input(_emptyInput)
				}
				return _cState;
			}

			/**
			 * Sets the start state.
			 * @param id The ID of an existing state
			 * @private
			 */
			function _setStartState(id) {
				_startStateId = id;
			}

			/**
			 * Gets the current state(s).
			 * @returns {Array} The IDs of the current state(s)
			 * @private
			 */
			function _getCurrentStates() {
				return _cState;
			}

			/**
			 * Determines if the current state is a final state.
			 * @returns {boolean} Is current state a final state?
			 * @private
			 */
			function _isAccepted() {
				for(var i in _cState) {
					//noinspection JSUnfilteredForInLoop
					if(_states[_cState[i]].final)
						return true;
				}
				return false;
			}



			/* =========== *
			 * Constructor *
			 * =========== */

			(function(states, startState) {
				// add the default null state
				_addState(_nullStateId, false, []);

				if(states !== undefined)
					for(var stateName in states) { //noinspection JSUnfilteredForInLoop
						_addState(stateName, states[stateName].final, states[stateName].transitions);
					}

				// Make the explicity-declared start state be startState
				if(startState !== undefined)
					_setStartState(startState);

				// Run the machine
				_reset();
			})(states, startState);



			/* ================== *
			 * Privileged Members *
			 * ================== */

			return {
				/**
				 * Resets the machine.
				 * @returns {Array} The current state
				 */
				reset: function() {
					return _reset();
				},

				/**
				 * Inputs a symbol to the machine.
				 * @param i The input symbol
				 * @returns {Array} The resulting state
				 */
				input: function(i) {
					return _input(i);
				},

				/**
				 * Gets the current state(s).
				 * @returns {Array} The IDs of the current state(s)
				 */
				state: function() {
					return _getCurrentStates();
				},

				/**
				 * Determines if the current state is a final state.
				 * @returns {boolean} Is current state a final state?
				 */
				accepted: function() {
					return _isAccepted();
				},
				
				nullState: function() {
					return _getCurrentStates().length == 0;
				}
			};
		}
	};
})(window);


src/avtomat-1.1.js
/**                                                                                                                  *
 *  _______________________________________________________________________________________________________________  *
 *                                                                                                                   *
 *    AVTOMAT - Non-deterministic Finite State Machine (with Empty Moves) Implementation                             *
 *  _______________________________________________________________________________________________________________  *
 *                                                                                                                   *
 *    Version: 1.1                                                                                                   *
 *    Author:  Temoto-kun                                                                                            *
 *    Date:    2013 December 13   08:50 [UTC+8]                                                                      *
 *                                                                                                                   *
 *  =============================================================================================================== **/

(function(w) {
	w.Avtomat = w.Avtomat || {
		/**
		 * A non-deterministic finite state machine which supports empty moves for transitions.
		 * @param states JSON object containing the states, and their details (is final state, transitions).
		 * @param startState The ID of the start state
		 * @constructor
		 */
		StateMachine : function(states, startState) {



			/* =============== *
			 * Private Members *
			 * =============== */

			/* ------ *
			 * Fields *
			 * ------ */

 			var
				// The states
				_states = [],

				// The ID for the "null" state
				_nullStateId = null,

				// The string for the empty input
				_emptyInput = "",

				// IDs of the current state
				_cState = [_nullStateId],

				// ID of the start state
				_startStateId = startState || _nullStateId,

		        // events triggered...
			    _events = {
				    // ... during state change
					changing: [],
				    // after state change
					change: []
				}
				;

			/* ------- *
			 * Methods *
			 * ------- */

 			/**
			 * Adds a transition from one state to another.
			 * @param idFrom The ID of the source state
			 * @param input The input symbol
			 * @param idTo The ID of the destination state
			 * @private
			 */
			function _addTransition(idFrom, input, idTo) {
				_states[idFrom].transitions[input] = idTo;
			}

			function _deleteTransition(id, input) {
				// FIXME dunno what this does
				// delete _states[idFrom].transitions[input];
				_states[idFrom].transitions[input] = undefined;
			}

			/**
			 * Adds a new state.
			 * @param id The ID of the new state
			 * @param isFinal Is state a final state?
			 * @param transitions Associative array containing the input symbol, and the destination state id
			 * @see _createBlankState
			 * @see _addTransition
			 * @private
			 */
			function _addState(id, isFinal, transitions) {
				if(_states[id] === undefined) {
					if(_startStateId == _nullStateId)
						// make the first non-null state to be startState, given that there
						_setStartState(id);

					_states[id] = _createBlankState();

					if(id != _nullStateId) {
						_setFinal(id, isFinal);
						if(transitions !== undefined)
							for(var input in transitions)
								_addTransition(id, input, transitions[input]);
					}
				}
			}

			function _deleteState(id) {
				// FIXME dunno what this does
				// delete _states[id];
				_states[id] = undefined;
			}

			/**
			 * Creates a prefab blank state.
			 * @returns {{final: boolean, transitions: Array}} New blank non-final state with no transitions
			 * @private
			 */
			function _createBlankState() {
				return {
					final: false,
					transitions: [],
					events: {
						arriving: [],
						arrive: [],
						leaving: [],
						leave: []
					}
				};
			}

			/**
			 * Sets a state to be final state.
			 * @param id The ID of the state
			 * @param bFinal Is state a final state?
			 * @private
			 */
			function _setFinal(id, bFinal) {
				if(id != _nullStateId)
					_states[id].final = bFinal;
			}

			/**
			 * Resets the machine.
			 * @returns {Array} The current state
			 * @private
			 */
			function _reset() {
				_cState = [_startStateId];
				_input("");
				return _cState;
			}

			/**
			 * Inputs a symbol to the machine.
			 * @param input The input symbol
			 * @returns {Array} The resulting state
			 * @private
			 */
			function _input(input) {
				var
					// Queue for old states
					srcStates = _cState,

					// Stack for new states
					newStates = [],

					// If the new states have empty moves
					haveEmptyMoves = false,

					events;

				// Trigger machine state changing
				events = _events.changing;

				// Trigger leaving events
				for(var e in events)
					events[e]();

				/*
				 * _input operates with a queue as container for the state IDs for processing. It empties srcStates to
				 * make sure all the states will receive the input.
				 */
				while(srcStates.length > 0) {
					// Dequeue
					var oldState = srcStates.shift();

					events = _states[oldState].events.leaving;

					// Trigger leaving events
					for(e in events)
						events[e]();

					// Destination states (non-array), i.e. state ID strings by themselves
					var destStateObjs = _states[oldState].transitions[input]
						// if there is no explicity-defined destination states for empty moves, current state is
						// retained
						|| (input == _emptyInput ? oldState : _nullStateId);
					// Destination states (array)
					var destStates;

					// Destination states must be turned into an array for processing
					if(destStateObjs instanceof Array)
						destStates = destStateObjs;
					else
						destStates = [destStateObjs];

					// Determine the destination states
					for(var j = 0; j < destStates.length; j++) {
						var containsState = false;
						for(var k = 0, cStateLen = newStates.length; k < cStateLen; k++)
							containsState = (containsState || (newStates[k] == destStates[j]));
						if(!containsState && destStates[j] != _nullStateId) {
							// Make sure the destination states don't duplicate. Null states are not included as well.

							// Trigger leave events
							events = _states[oldState].events.leave;
							for(e in events)
								events[e]();

							// Trigger arriving events
							events = _states[destStates[j]].events.arriving;
							for(e in events)
								events[e]();

							newStates.push(destStates[j]);

							// Trigger arrive events
							events = _states[destStates[j]].events.arrive;
							for(e in events)
								events[e]();
						}
					}

					// Terminate the state transitions when there are no explicitly-defined empty transitions
					if(input == _emptyInput && destStateObjs == _nullStateId)
						haveEmptyMoves = false;
				}
				_cState = newStates;

				// Trigger machine state changing
				events = _events.change;

				// Trigger leaving events
				for(e in events)
					events[e]();

				// can't have empty moves when machine doesn't have current states
				if(newStates.length > 0) {
					for(var l in newStates) {
						if(_states[newStates[l]].transitions[_emptyInput] !== undefined) {
							haveEmptyMoves = true;
							break;
						}
					}
					if(haveEmptyMoves)
						_input(_emptyInput)
				}
				return _cState;
			}

			/**
			 * Sets the start state.
			 * @param id The ID of an existing state
			 * @private
			 */
			function _setStartState(id) {
				_startStateId = id;
			}

			/**
			 * Gets the current state(s).
			 * @returns {Array} The IDs of the current state(s)
			 * @private
			 */
			function _getCurrentStates() {
				return _cState;
			}

			/**
			 * Determines if the current state is a final state.
			 * @returns {boolean} Is current state a final state?
			 * @private
			 */
			function _isAccepted() {
				for(var i in _cState)
					if(_states[_cState[i]]["final"])
						return true;
				return false;
			}

			function _bindStateEvent(stateId, type, fn) {
				_states[stateId].events[type].push(fn);
			}

			function _bindMachineEvent(type, fn) {
				_events[type].push(fn);
			}

			/*
			function _hasTransitions(id) {
				if(id == undefined)
					id = _getCurrentStates();
				else if(!(id instanceof Array))
					id = [id];

				var b = false;
				for(var state in id) {
					var t = 0;
					for(var transitions in _states[id].transitions)
						t++;
					b = b || t > 0;
				}
				return b;
			}
			*/

			function _hasTransition(id, input) {
				return _states[id].transitions[input] !== undefined || input == _emptyInput;
			}

			function _hasTransitions(id) {
				for(var transitions in _states[id].transitions)
					return true;
				return false;
			}



			/* =========== *
			 * Constructor *
			 * =========== */

			(function(states, startState) {
				// add the default null state
				_addState(_nullStateId, false, []);

				if(states !== undefined)
					for(var stateName in states)
						_addState(stateName, states[stateName]["final"], states[stateName].transitions);

				// Make the explicity-declared start state be startState
				if(startState !== undefined)
					_setStartState(startState);

				// Run the machine
				_reset();
			})(states, startState);



			/* ================== *
			 * Privileged Members *
			 * ================== */

			return {
				/**
				 * Resets the machine.
				 * @returns {Array} The current state
				 */
				reset: _reset,

				/**
				 * Inputs a symbol to the machine.
				 * @param i The input symbol
				 * @returns {Array} The resulting state
				 */
				input: _input,

				/**
				 * Gets the current state(s).
				 * @returns {Array} The IDs of the current state(s)
				 */
				state: _getCurrentStates,

				/**
				 * Determines if the current state is a final state.
				 * @returns {boolean} Is current state a final state?
				 */
				accepted: _isAccepted,
				
				nullState: function() {
					return _getCurrentStates().length == 0;
				},

				addState: _addState,

				deleteState: _deleteState,

				addTransition: _addTransition,

				deleteTransition: _deleteTransition,

				hasTransition: _hasTransition,

				hasTransitions: function(id) {
					if(id == undefined)
						id = _getCurrentStates();
					else if(id instanceof Array) {
						var b = false;
						for(var state in id)
							b = b || _hasTransitions[state];
						return b;
					}
					return typeof(id) == "string" && _hasTransitions(id);
				},

				bindStateEvent: _bindStateEvent,

				bindMachineEvent: _bindMachineEvent
			};
		}
	};
})(window);
